client/tx/decode.go

```diff
@@ -0,0 +1,104 @@
+package tx
+
+import (
+	"encoding/base64"
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/types/rest"
+	"github.com/spf13/cobra"
+	"github.com/tendermint/go-amino"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/client/context"
+	"github.com/cosmos/cosmos-sdk/x/auth"
+)
+
+type (
+	// DecodeReq defines a tx decoding request.
+	DecodeReq struct {
+		Tx string `json:"tx"`
+	}
+
+	// DecodeResp defines a tx decoding response.
+	DecodeResp struct {
+		Tx auth.StdTx `json:"tx"`
+	}
+)
+
+// DecodeTxRequestHandlerFn returns the decode tx REST handler. In particular,
+// it takes base64-decoded bytes, decodes it from the Amino wire protocol,
+// and responds with a json-formatted transaction.
+func DecodeTxRequestHandlerFn(cdc *codec.Codec, cliCtx context.CLIContext) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var req DecodeReq
+
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
+			return
+		}
+
+		err = cdc.UnmarshalJSON(body, &req)
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
+			return
+		}
+
+		txBytes, err := base64.StdEncoding.DecodeString(req.Tx)
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
+			return
+		}
+
+		var stdTx auth.StdTx
+		err = cliCtx.Codec.UnmarshalBinaryLengthPrefixed(txBytes, &stdTx)
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
+			return
+		}
+
+		response := DecodeResp{Tx: stdTx}
```

Comment:
In the same spirit of `Encode`, why not just return the `StdTx directly? I don't think there's a need to nest it under `tx`.

Reply:
Great idea !  Done!

Reply:
wapper `StdTx` in `tx` would make `swagger-ui.yml` more clean and grace.

ref: https://github.com/cosmos/cosmos-sdk/issues/4141#issuecomment-484312785

Reply:
@yangyanqing for encode I get that, but I don't get why we need to do this for a decoded response. There is no need to nest.